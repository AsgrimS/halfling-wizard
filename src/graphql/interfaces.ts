import {
  Class,
  Subclass,
  Proficiency,
  ClassProficiency_Choices,
  AbilityScoreSkills,
  ClassSpellcasting,
  ClassStarting_Equipment,
} from "./schema";

export interface HeroClassesData {
  classes: Class[];
}

export interface HeroClassData {
  class: Class;
}

export interface SubclassesData {
  subclasses: Subclass[];
}

export interface ProficienciesData {
  proficiencies: Proficiency[];
}

export interface ProficienciesChoicesData {
  proficienciesChoices: ClassProficiency_Choices[];
}
export interface SavingThrowsData {
  saving_throws: AbilityScoreSkills[];
}

export interface ClassSpellcastingData {
  spellcasting: ClassSpellcasting;
}

export interface StartingEquipmentData {
  startingEquipments: ClassStarting_Equipment[];
}
