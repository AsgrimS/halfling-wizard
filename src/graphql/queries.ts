import { gql } from "@apollo/client";

export const HERO_CLASS = gql`
  query GetHeroesClasses($classIndex: String!) {
    class(filter: { name: $classIndex }) {
      hit_die
      index
      name
      proficiencies {
        index
        name
      }
      saving_throws {
        index
        name
      }
      proficiency_choices {
        choose
        from {
          name
          index
        }
      }
      spellcasting {
        level
        spellcasting_ability {
          index
          name
        }
      }
      starting_equipment {
        equipment {
          name
          index
        }
        quantity
      }
      starting_equipment_options {
        choose
        from {
          equipment {
            name
            index
          }
          quantity
        }
      }
      subclasses {
        index
        name
      }
    }
  }
`;

export const HEROES_CLASSES = gql`
  query GetHeroesClasses {
    classes {
      name
      index
    }
  }
`;
