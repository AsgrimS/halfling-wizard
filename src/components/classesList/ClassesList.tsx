import React from "react";

import { useQuery } from "@apollo/client";
import styled from "styled-components";

import { HEROES_CLASSES } from "../../graphql/queries";
import { HeroClassesData } from "../../graphql/interfaces";
import { LazyQueryCallbackFunction } from "../../utils/interfaces";
import ExpandBtn from "../../styled-components/ExpandBtn";
import LoadingAnimation from "../../styled-components/LoadingAnimation";

const ClassesList: React.FC<LazyQueryCallbackFunction> = ({
  getClassDetail,
}) => {
  const heroClassesData = useQuery<HeroClassesData>(HEROES_CLASSES);

  return (
    <ListContainer>
      <ListTitle>Classes</ListTitle>
      <List>
        {heroClassesData.loading ? (
          <LoadingAnimation />
        ) : (
          <>
            {heroClassesData.data &&
              heroClassesData.data.classes.map((heroClass) => (
                <li>
                  <ExpandBtn
                    type="button"
                    onClick={() =>
                      getClassDetail({
                        variables: { classIndex: heroClass.name },
                      })
                    }
                  >
                    {heroClass.name}
                  </ExpandBtn>
                </li>
              ))}
          </>
        )}
      </List>
    </ListContainer>
  );
};

export default ClassesList;

const ListContainer = styled.div`
  width: 100%;
  grid-column-start: 1;
  grid-column-end: 1;
  background-color: var(--blackOliveL);
  border-radius: 0.5rem;
  align-self: start;
`;

const ListTitle = styled.span`
  font-family: "PT Sans Bold";
  font-size: 1.5rem;
  margin-left: 0.5rem;
`;

const List = styled.ul`
  padding-left: 1rem;
  list-style-type: none;
`;
