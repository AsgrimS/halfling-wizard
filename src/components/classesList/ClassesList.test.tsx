import React from "react";
import { act } from "react-dom/test-utils";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

import ClassesList from "./ClassesList";
import { Class } from "../../graphql/schema";
import { HEROES_CLASSES } from "../../graphql/queries";

describe("ClassesList component", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render a list of buttons containing avaible classes", async () => {
    const mockClassesListData: Class[] = [
      { name: "Wizard", index: "wizard" },
      { name: "Barbarian", index: "barbarian" },
    ];
    const mockGetClassDetail = jest.fn();
    const mockQuery: MockedResponse = {
      request: { query: HEROES_CLASSES },
      result: { data: { classes: mockClassesListData } },
    };

    act(() => {
      render(
        <MockedProvider mocks={[mockQuery]} addTypename={false}>
          <ClassesList getClassDetail={mockGetClassDetail} />
        </MockedProvider>
      );
    });
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    const subclasessListElement = screen.getByTestId("classes-list");
    const btnElements = subclasessListElement.getElementsByTagName("button");

    expect(subclasessListElement).toBeInTheDocument();
    expect(btnElements.length).toEqual(2);

    for (let i = 0; i < btnElements.length; i += 1) {
      expect(btnElements[i]).toHaveTextContent(mockClassesListData[i].name!);
    }

    btnElements[0].dispatchEvent(new MouseEvent("click", { bubbles: true }));
    fireEvent.click(btnElements[0]);
    expect(mockGetClassDetail).toBeCalledTimes(2);

    expect(subclasessListElement).toMatchSnapshot();
  });
});
