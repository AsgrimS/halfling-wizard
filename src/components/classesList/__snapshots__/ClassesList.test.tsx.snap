// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassesList component should render a list of buttons containing avaible classes 1`] = `
<section
  class="sc-hKFxyN sc-eCApnc jJwpmv dvmDAx"
  data-testid="classes-list"
>
  <span
    class="sc-jSFjdj fWvYUO"
  >
    Classes
  </span>
  <ul
    class="sc-dlnjwi sc-gKAaRy jnsiJj idWzYF"
  >
    <li
      class="sc-iCoGMd cUbQsk"
    >
      <button
        class="sc-bdnxRM fFtxCn"
        type="button"
      >
        Wizard
      </button>
    </li>
    <li
      class="sc-iCoGMd cUbQsk"
    >
      <button
        class="sc-bdnxRM fFtxCn"
        type="button"
      >
        Barbarian
      </button>
    </li>
  </ul>
</section>
`;
